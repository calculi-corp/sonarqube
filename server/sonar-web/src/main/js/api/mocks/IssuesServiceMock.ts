/*
 * SonarQube
 * Copyright (C) 2009-2022 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { cloneDeep, keyBy, range, times } from 'lodash';
import { RuleDescriptionSections } from '../../apps/coding-rules/rule';
import {
  mockSnippetsByComponent,
  mockSourceLine,
  mockSourceViewerFile
} from '../../helpers/mocks/sources';
import { RequestData } from '../../helpers/request';
import { getStandards } from '../../helpers/security-standard';
import { mockPaging, mockRawIssue, mockRuleDetails } from '../../helpers/testMocks';
import { BranchParameters } from '../../types/branch-like';
import { RawFacet, RawIssue, RawIssuesResponse, ReferencedComponent } from '../../types/issues';
import { Standards } from '../../types/security';
import {
  Dict,
  RuleActivation,
  RuleDetails,
  SnippetsByComponent,
  SourceViewerFile
} from '../../types/types';
import { getComponentForSourceViewer, getSources } from '../components';
import { getIssueFlowSnippets, searchIssues } from '../issues';
import { getRuleDetails } from '../rules';

function mockReferenceComponent(override?: Partial<ReferencedComponent>) {
  return {
    key: 'component1',
    name: 'Component1',
    uuid: 'id1',
    ...override
  };
}

interface IssueData {
  issue: RawIssue;
  snippets: Dict<SnippetsByComponent>;
}

export default class IssuesServiceMock {
  isAdmin = false;
  standards?: Standards;
  sourceViewerFiles: SourceViewerFile[];
  list: IssueData[];

  constructor() {
    this.sourceViewerFiles = [
      mockSourceViewerFile('file.foo', 'project'),
      mockSourceViewerFile('file.bar', 'project')
    ];
    this.list = [
      {
        issue: mockRawIssue(false, {
          key: 'issue0',
          component: 'project:file.foo',
          message: 'Issue on file',
          rule: 'simpleRuleId',
          textRange: undefined,
          line: undefined
        }),
        snippets: {}
      },
      {
        issue: mockRawIssue(false, {
          key: 'issue1',
          component: 'project:file.foo',
          message: 'Fix this',
          rule: 'simpleRuleId',
          textRange: {
            startLine: 10,
            endLine: 10,
            startOffset: 0,
            endOffset: 2
          },
          flows: [
            {
              locations: [
                {
                  component: 'project:file.foo',
                  textRange: {
                    startLine: 1,
                    endLine: 1,
                    startOffset: 0,
                    endOffset: 1
                  }
                }
              ]
            },
            {
              locations: [
                {
                  component: 'project:file.bar',
                  textRange: {
                    startLine: 20,
                    endLine: 20,
                    startOffset: 0,
                    endOffset: 1
                  }
                }
              ]
            }
          ]
        }),
        snippets: keyBy(
          [
            mockSnippetsByComponent(
              'file.foo',
              'project',
              times(40, i => i + 1)
            ),
            mockSnippetsByComponent(
              'file.bar',
              'project',
              times(40, i => i + 1)
            )
          ],
          'component.key'
        )
      },
      {
        issue: mockRawIssue(false, {
          key: 'issue2',
          component: 'project:file.bar',
          message: 'Fix that',
          rule: 'advancedRuleId',
          textRange: {
            startLine: 25,
            endLine: 25,
            startOffset: 0,
            endOffset: 1
          }
        }),
        snippets: keyBy(
          [
            mockSnippetsByComponent(
              'file.bar',
              'project',
              times(40, i => i + 20)
            )
          ],
          'component.key'
        )
      },
      {
        issue: mockRawIssue(false, {
          key: 'issue3',
          component: 'project:file.bar',
          message: 'Second issue',
          rule: 'other',
          textRange: {
            startLine: 28,
            endLine: 28,
            startOffset: 0,
            endOffset: 1
          }
        }),
        snippets: keyBy(
          [
            mockSnippetsByComponent(
              'file.bar',
              'project',
              times(40, i => i + 20)
            )
          ],
          'component.key'
        )
      }
    ];
    (searchIssues as jest.Mock).mockImplementation(this.handleSearchIssues);
    (getRuleDetails as jest.Mock).mockImplementation(this.handleGetRuleDetails);
    (getIssueFlowSnippets as jest.Mock).mockImplementation(this.handleGetIssueFlowSnippets);
    (getSources as jest.Mock).mockImplementation(this.handleGetSources);
    (getComponentForSourceViewer as jest.Mock).mockImplementation(
      this.handleGetComponentForSourceViewer
    );
  }

  async getStandards(): Promise<Standards> {
    if (this.standards) {
      return this.standards;
    }
    this.standards = await getStandards();
    return this.standards;
  }

  owasp2021FacetList(): RawFacet {
    return {
      property: 'owaspTop10-2021',
      values: [{ val: 'a1', count: 0 }]
    };
  }

  setIsAdmin(isAdmin: boolean) {
    this.isAdmin = isAdmin;
  }

  handleGetSources = (data: { key: string; from?: number; to?: number } & BranchParameters) => {
    return this.reply(range(data.from || 1, data.to || 10).map(line => mockSourceLine({ line })));
  };

  handleGetComponentForSourceViewer = (data: { component: string } & BranchParameters) => {
    const file = this.sourceViewerFiles.find(f => f.key === data.component);
    if (file === undefined) {
      return Promise.reject({
        errors: [{ msg: `No source file has been found for id ${data.component}` }]
      });
    }

    return this.reply(file);
  };

  handleGetIssueFlowSnippets = (issueKey: string): Promise<Dict<SnippetsByComponent>> => {
    const issue = this.list.find(i => i.issue.key === issueKey);
    if (issue === undefined) {
      return Promise.reject({ errors: [{ msg: `No issue has been found for id ${issueKey}` }] });
    }
    return this.reply(issue.snippets);
  };

  handleGetRuleDetails = (parameters: {
    actives?: boolean;
    key: string;
  }): Promise<{ actives?: RuleActivation[]; rule: RuleDetails }> => {
    if (parameters.key === 'advancedRuleId') {
      return this.reply({
        rule: mockRuleDetails({
          key: parameters.key,
          name: 'Advanced rule',
          descriptionSections: [
            { key: RuleDescriptionSections.INTRODUCTION, content: '<h1>Into</h1>' },
            { key: RuleDescriptionSections.ROOT_CAUSE, content: '<h1>Because</h1>' },
            { key: RuleDescriptionSections.HOW_TO_FIX, content: '<h1>Fix with</h1>' },
            {
              content: '<p> Context 1 content<p>',
              key: RuleDescriptionSections.HOW_TO_FIX,
              context: {
                displayName: 'Spring'
              }
            },
            {
              content: '<p> Context 2 content<p>',
              key: RuleDescriptionSections.HOW_TO_FIX,
              context: {
                displayName: 'Context 2'
              }
            },
            {
              content: '<p> Context 3 content<p>',
              key: RuleDescriptionSections.HOW_TO_FIX,
              context: {
                displayName: 'Context 3'
              }
            },
            { key: RuleDescriptionSections.RESOURCES, content: '<h1>Link</h1>' }
          ]
        })
      });
    }
    return this.reply({
      rule: mockRuleDetails({
        key: parameters.key,
        name: 'Simple rule',
        htmlNote: '<h1>Note</h1>',
        descriptionSections: [
          {
            key: RuleDescriptionSections.DEFAULT,
            content: '<h1>Default</h1> Default description'
          }
        ]
      })
    });
  };

  handleSearchIssues = (query: RequestData): Promise<RawIssuesResponse> => {
    const facets = (query.facets ?? '').split(',').map((name: string) => {
      if (name === 'owaspTop10-2021') {
        return this.owasp2021FacetList();
      }
      return {
        property: name,
        values: []
      };
    });
    return this.reply({
      components: [mockReferenceComponent()],
      effortTotal: 199629,
      facets,
      issues: this.list.map(line => line.issue),
      languages: [],
      paging: mockPaging()
    });
  };

  reply<T>(response: T): Promise<T> {
    return Promise.resolve(cloneDeep(response));
  }
}
