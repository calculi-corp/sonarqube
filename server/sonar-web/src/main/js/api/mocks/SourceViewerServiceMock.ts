/*
 * SonarQube
 * Copyright (C) 2009-2022 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
/* eslint-disable no-template-curly-in-string */

import { cloneDeep, pick, times } from 'lodash';
import {
  getComponentData,
  getComponentForSourceViewer,
  getDuplications,
  getSources
} from '../../api/components';
import { mockSourceLine } from '../../helpers/mocks/sources';
import { BranchParameters } from '../../types/branch-like';
import { Dict } from '../../types/types';

function mockSourceFileView(name: string) {
  return {
    component: {
      key: `project:${name}`,
      name,
      qualifier: 'FIL',
      path: name,
      language: 'js',
      analysisDate: '2019-08-08T12:15:12+0200',
      leakPeriodDate: '2018-08-07T11:22:22+0200',
      version: '1.2-SNAPSHOT',
      needIssueSync: false
    },
    sourceFileView: {
      key: `project:${name}`,
      uuid: 'AWMgNpveti8CNlpVyHAm',
      path: name,
      name,
      longName: name,
      q: 'FIL',
      project: 'project',
      projectName: 'Test project',
      fav: false,
      canMarkAsFavorite: true,
      measures: { lines: '0', issues: '0' }
    }
  };
}

const ANCESTORS = [
  {
    key: 'project',
    name: 'Test project',
    description: '',
    qualifier: 'TRK',
    analysisDate: '2019-08-08T12:15:12+0200',
    tags: [],
    visibility: 'public',
    leakPeriodDate: '2018-08-07T11:22:22+0200',
    version: '1.2-SNAPSHOT',
    needIssueSync: false
  }
];
const FILES: Dict<any> = {
  'project:test3.js': {
    ...mockSourceFileView('test3.js'),
    sources: [],
    ancestors: ANCESTORS
  },
  'project:test2.js': {
    ...mockSourceFileView('test2.js'),
    sources: times(200, n =>
      mockSourceLine({
        line: n,
        code: `\u003cspan class\u003d"cd"\u003eLine ${n}\u003c/span\u003e`
      })
    ),
    ancestors: ANCESTORS
  },
  'project:test.js': {
    ...mockSourceFileView('test.js'),
    sources: [
      {
        line: 1,
        code: '\u003cspan class\u003d"cd"\u003e/*\u003c/span\u003e',
        scmRevision: 'f09ee6b610528aa37b7b51be395c93524cebae8f',
        scmAuthor: 'stas.vilchik@sonarsource.com',
        scmDate: '2018-07-10T20:21:20+0200',
        duplicated: false,
        isNew: false,
        lineHits: 1,
        coveredConditions: 1
      },
      {
        line: 2,
        code: '\u003cspan class\u003d"cd"\u003e * SonarQube\u003c/span\u003e',
        scmRevision: 'f09ee6b610528aa37b7b51be395c93524cebae8f',
        scmAuthor: 'stas.vilchik@sonarsource.com',
        scmDate: '2018-07-10T20:21:20+0200',
        duplicated: false,
        isNew: false,
        lineHits: 0,
        conditions: 1
      },
      {
        line: 3,
        code: '\u003cspan class\u003d"cd"\u003e * Copyright\u003c/span\u003e',
        scmRevision: '89a3d21bc28f2fa6201b5e8b1185d5358481b3dd',
        scmAuthor: 'pierre.guillot@sonarsource.com',
        scmDate: '2022-01-28T21:03:07+0100',
        duplicated: false,
        isNew: false,
        lineHits: 1
      },
      {
        line: 4,
        code:
          '\u003cspan class\u003d"cd"\u003e * mailto:info AT sonarsource DOT com\u003c/span\u003e',
        scmRevision: 'f09ee6b610528aa37b7b51be395c93524cebae8f',
        scmAuthor: 'stas.vilchik@sonarsource.com',
        duplicated: false,
        isNew: false,
        lineHits: 1,
        conditions: 1,
        coveredConditions: 1
      },
      {
        line: 5,
        code: '\u003cspan class\u003d"cd"\u003e * 5\u003c/span\u003e',
        scmRevision: 'f04ee6b610528aa37b7b51be395c93524cebae8f',
        duplicated: false,
        isNew: false,
        lineHits: 2,
        conditions: 2,
        coveredConditions: 1
      },
      {
        line: 6,
        code: '\u003cspan class\u003d"cd"\u003e * 6\u003c/span\u003e',
        scmRevision: 'f04ee6b610528aa37b7b51be395c93524cebae8f',
        duplicated: false,
        isNew: false,
        lineHits: 0
      },
      {
        line: 7,
        code: '\u003cspan class\u003d"cd"\u003e * 7\u003c/span\u003e',
        scmRevision: 'f04ee6b610528aa37b7b51be395c93524cebae8f',
        duplicated: true,
        isNew: true
      },
      {
        code:
          '\u003cspan class\u003d"cd"\u003e * This program is free software; you can redistribute it and/or\u003c/span\u003e',
        scmRevision: 'f09ee6b610528aa37b7b51be395c93524cebae8f',
        scmAuthor: 'stas.vilchik@sonarsource.com',
        scmDate: '2018-07-10T20:21:20+0200',
        duplicated: false,
        isNew: false
      }
    ],
    ancestors: ANCESTORS,
    duplications: [
      {
        blocks: [
          { from: 7, size: 1, _ref: '1' },
          { from: 1, size: 1, _ref: '2' }
        ]
      }
    ],
    files: {
      '1': {
        key: 'project:test.js',
        name: 'test.js',
        uuid: 'AX8NSmj8EGYw5-dyy63J',
        project: 'project',
        projectUuid: 'AX7juKJqVeQLJMPyb_b-',
        projectName: 'project'
      },
      '2': {
        key: 'project:test2.js',
        name: 'test2.js',
        uuid: 'BX8NSmj8EGYw5-dyy63J',
        project: 'project',
        projectUuid: 'AX7juKJqVeQLJMPyb_b-',
        projectName: 'project'
      }
    }
  }
};

export class SourceViewerServiceMock {
  constructor() {
    (getComponentData as jest.Mock).mockImplementation(this.handleGetComponentData);
    (getComponentForSourceViewer as jest.Mock).mockImplementation(
      this.handleGetComponentForSourceViewer
    );
    (getDuplications as jest.Mock).mockImplementation(this.handleGetDuplications);
    (getSources as jest.Mock).mockImplementation(this.handleGetSources);
  }

  getHugeFile(): string {
    return 'project:test2.js';
  }

  getEmptyFile(): string {
    return 'project:test3.js';
  }

  getFileWithSource(): string {
    return 'project:test.js';
  }

  handleGetSources = (data: { key: string; from?: number; to?: number } & BranchParameters) => {
    const { sources } = FILES[data.key];
    const from = data.from || 1;
    const to = data.to || sources.length;
    return this.reply(sources.slice(from - 1, to));
  };

  handleGetDuplications = (data: { key: string } & BranchParameters) => {
    return this.reply(pick(FILES[data.key], ['duplications', 'files']));
  };

  handleGetComponentForSourceViewer = (data: { component: string } & BranchParameters) => {
    return this.reply(FILES[data.component]['sourceFileView']);
  };

  handleGetComponentData = (data: { component: string } & BranchParameters) => {
    return this.reply(pick(FILES[data.component], ['component', 'ancestor']));
  };

  reply<T>(response: T): Promise<T> {
    return Promise.resolve(cloneDeep(response));
  }
}
