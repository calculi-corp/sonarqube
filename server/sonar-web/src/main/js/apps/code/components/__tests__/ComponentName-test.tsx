/*
 * SonarQube
 * Copyright (C) 2009-2022 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { shallow } from 'enzyme';
import * as React from 'react';
import { mockMainBranch } from '../../../../helpers/mocks/branch-like';
import { mockComponentMeasure } from '../../../../helpers/mocks/component';
import { ComponentQualifier } from '../../../../types/component';
import ComponentName, { getTooltip, Props } from '../ComponentName';

describe('#getTooltip', () => {
  it('should correctly format component information', () => {
    expect(getTooltip(mockComponentMeasure(true))).toMatchSnapshot();
    expect(getTooltip(mockComponentMeasure(true, { qualifier: 'UTS' }))).toMatchSnapshot();
    expect(getTooltip(mockComponentMeasure(true, { path: undefined }))).toMatchSnapshot();
    expect(getTooltip(mockComponentMeasure(false))).toMatchSnapshot();
  });
});

describe('#ComponentName', () => {
  it('should render correctly for files', () => {
    expect(shallowRender()).toMatchSnapshot();
    expect(shallowRender({ canBrowse: true })).toMatchSnapshot();
    expect(
      shallowRender({ rootComponent: mockComponentMeasure(false, { qualifier: 'TRK' }) })
    ).toMatchSnapshot();
    expect(
      shallowRender({ rootComponent: mockComponentMeasure(false, { qualifier: 'APP' }) })
    ).toMatchSnapshot();
    expect(
      shallowRender({
        component: mockComponentMeasure(true, { branch: 'foo' }),
        rootComponent: mockComponentMeasure(false, { qualifier: 'APP' })
      })
    ).toMatchSnapshot();
    expect(shallowRender({ newCodeSelected: true })).toMatchSnapshot();
    expect(shallowRender({ newCodeSelected: false })).toMatchSnapshot();
  });

  it('should render correctly for dirs', () => {
    expect(
      shallowRender({
        component: mockComponentMeasure(false, { name: 'src/main/ts/app', qualifier: 'DIR' }),
        previous: mockComponentMeasure(false, { name: 'src/main/ts/tests', qualifier: 'DIR' })
      })
    ).toMatchSnapshot();
    expect(
      shallowRender({
        component: mockComponentMeasure(false, { name: 'src', qualifier: 'DIR' }),
        previous: mockComponentMeasure(false, { name: 'lib', qualifier: 'DIR' })
      })
    ).toMatchSnapshot();
  });

  it('should render correctly for refs', () => {
    expect(
      shallowRender({
        component: mockComponentMeasure(false, {
          branch: 'foo',
          refKey: 'src/main/ts/app',
          qualifier: ComponentQualifier.Project
        })
      })
    ).toMatchSnapshot();
    expect(
      shallowRender({
        component: mockComponentMeasure(false, {
          branch: 'foo',
          refKey: 'src/main/ts/app',
          qualifier: ComponentQualifier.Project
        }),
        rootComponent: mockComponentMeasure(false, { qualifier: ComponentQualifier.Application })
      })
    ).toMatchSnapshot();

    expect(
      shallowRender({
        component: mockComponentMeasure(false, {
          refKey: 'src/main/ts/app',
          qualifier: ComponentQualifier.Project
        }),
        rootComponent: mockComponentMeasure(false, { qualifier: ComponentQualifier.Portfolio })
      })
    ).toMatchSnapshot();
  });

  it.each([
    [ComponentQualifier.Application, 'refKey'],
    [ComponentQualifier.Portfolio, 'refKey'],
    [ComponentQualifier.SubPortfolio, 'refKey'],
    [ComponentQualifier.Project, 'refKey'],
    [ComponentQualifier.Project, undefined]
  ])('should render breadcrumb correctly for %s', (qualifier, refKey) => {
    expect(
      shallowRender({
        component: mockComponentMeasure(false, {
          refKey,
          qualifier
        }),
        unclickable: true
      })
    ).toMatchSnapshot();
  });
});

function shallowRender(props: Partial<Props> = {}) {
  return shallow(
    <ComponentName
      branchLike={mockMainBranch()}
      component={mockComponentMeasure(true)}
      rootComponent={mockComponentMeasure()}
      {...props}
    />
  );
}
