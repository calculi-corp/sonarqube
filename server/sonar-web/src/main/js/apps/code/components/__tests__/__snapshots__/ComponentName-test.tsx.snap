// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#ComponentName should render breadcrumb correctly for APP 1`] = `
<span
  aria-label="code.parent_folder"
  className="max-width-100 display-inline-block text-ellipsis"
  title="Foo

foo"
>
  <span>
    <QualifierIcon
      qualifier="APP"
    />
     
    Foo
  </span>
</span>
`;

exports[`#ComponentName should render breadcrumb correctly for SVW 1`] = `
<span
  aria-label="code.parent_folder"
  className="max-width-100 display-inline-block text-ellipsis"
  title="Foo

foo"
>
  <span>
    <QualifierIcon
      qualifier="SVW"
    />
     
    Foo
  </span>
</span>
`;

exports[`#ComponentName should render breadcrumb correctly for TRK 1`] = `
<span
  aria-label="code.parent_folder"
  className="max-width-100 display-inline-block text-ellipsis"
  title="Foo

foo"
>
  <span>
    <QualifierIcon
      qualifier="TRK"
    />
     
    Foo
  </span>
</span>
`;

exports[`#ComponentName should render breadcrumb correctly for TRK 2`] = `
<span
  aria-label="code.parent_folder"
  className="max-width-100 display-inline-block text-ellipsis"
  title="Foo

foo"
>
  <span>
    <QualifierIcon
      qualifier="TRK"
    />
     
    Foo
  </span>
</span>
`;

exports[`#ComponentName should render breadcrumb correctly for VW 1`] = `
<span
  aria-label="code.parent_folder"
  className="max-width-100 display-inline-block text-ellipsis"
  title="Foo

foo"
>
  <span>
    <QualifierIcon
      qualifier="VW"
    />
     
    Foo
  </span>
</span>
`;

exports[`#ComponentName should render correctly for dirs 1`] = `
<span
  className="max-width-100 display-inline-block text-ellipsis"
  title="src/main/ts/app

foo"
>
  <span>
    <QualifierIcon
      qualifier="DIR"
    />
     
    <span>
      <span
        style={
          Object {
            "color": "#666",
          }
        }
      >
        src/main/ts/
      </span>
      <span>
        app
      </span>
    </span>
  </span>
</span>
`;

exports[`#ComponentName should render correctly for dirs 2`] = `
<span
  className="max-width-100 display-inline-block text-ellipsis"
  title="src

foo"
>
  <span>
    <QualifierIcon
      qualifier="DIR"
    />
     
    src
  </span>
</span>
`;

exports[`#ComponentName should render correctly for files 1`] = `
<span
  className="max-width-100 display-inline-block text-ellipsis"
  title="src/index.tsx

foo:src/index.tsx"
>
  <span>
    <QualifierIcon
      qualifier="FIL"
    />
     
    index.tsx
  </span>
</span>
`;

exports[`#ComponentName should render correctly for files 2`] = `
<span
  className="max-width-100 display-inline-block text-ellipsis"
  title="src/index.tsx

foo:src/index.tsx"
>
  <Link
    className="link-with-icon"
    to={
      Object {
        "pathname": "/code",
        "search": "?id=foo&selected=foo%3Asrc%2Findex.tsx",
      }
    }
  >
    <QualifierIcon
      qualifier="FIL"
    />
     
    <span>
      index.tsx
    </span>
  </Link>
</span>
`;

exports[`#ComponentName should render correctly for files 3`] = `
<span
  className="max-width-100 display-inline-block text-ellipsis"
  title="src/index.tsx

foo:src/index.tsx"
>
  <span>
    <QualifierIcon
      qualifier="FIL"
    />
     
    index.tsx
  </span>
</span>
`;

exports[`#ComponentName should render correctly for files 4`] = `
<span
  className="max-width-100 display-inline-block text-ellipsis"
  title="src/index.tsx

foo:src/index.tsx"
>
  <span>
    <QualifierIcon
      qualifier="FIL"
    />
     
    index.tsx
  </span>
</span>
`;

exports[`#ComponentName should render correctly for files 5`] = `
<span
  className="max-width-100 display-inline-block text-ellipsis"
  title="src/index.tsx

foo:src/index.tsx"
>
  <span>
    <QualifierIcon
      qualifier="FIL"
    />
     
    index.tsx
  </span>
</span>
`;

exports[`#ComponentName should render correctly for files 6`] = `
<span
  className="max-width-100 display-inline-block text-ellipsis"
  title="src/index.tsx

foo:src/index.tsx"
>
  <span>
    <QualifierIcon
      qualifier="FIL"
    />
     
    index.tsx
  </span>
</span>
`;

exports[`#ComponentName should render correctly for files 7`] = `
<span
  className="max-width-100 display-inline-block text-ellipsis"
  title="src/index.tsx

foo:src/index.tsx"
>
  <span>
    <QualifierIcon
      qualifier="FIL"
    />
     
    index.tsx
  </span>
</span>
`;

exports[`#ComponentName should render correctly for refs 1`] = `
<span
  className="max-width-100 display-inline-block text-ellipsis"
  title="Foo

foo

foo"
>
  <Link
    className="link-with-icon"
    to={
      Object {
        "pathname": "/dashboard",
        "search": "?id=src%2Fmain%2Fts%2Fapp&code_scope=new",
      }
    }
  >
    <QualifierIcon
      qualifier="TRK"
    />
     
    <span>
      Foo
    </span>
  </Link>
</span>
`;

exports[`#ComponentName should render correctly for refs 2`] = `
<span
  className="max-width-100 display-inline-flex-center"
>
  <span
    className="text-ellipsis"
    title="Foo

foo

foo"
  >
    <Link
      className="link-with-icon"
      to={
        Object {
          "pathname": "/dashboard",
          "search": "?id=src%2Fmain%2Fts%2Fapp&branch=foo&code_scope=new",
        }
      }
    >
      <QualifierIcon
        qualifier="TRK"
      />
       
      <span>
        Foo
      </span>
    </Link>
  </span>
  <span
    className="text-ellipsis spacer-left"
  >
    <BranchIcon
      className="little-spacer-right"
    />
    <span
      className="note"
    >
      foo
    </span>
  </span>
</span>
`;

exports[`#ComponentName should render correctly for refs 3`] = `
<span
  className="max-width-100 display-inline-flex-center"
>
  <span
    className="text-ellipsis"
    title="Foo

foo"
  >
    <Link
      className="link-with-icon"
      to={
        Object {
          "pathname": "/dashboard",
          "search": "?id=src%2Fmain%2Fts%2Fapp&code_scope=new",
        }
      }
    >
      <QualifierIcon
        qualifier="TRK"
      />
       
      <span>
        Foo
      </span>
    </Link>
  </span>
  <span
    className="spacer-left badge flex-1"
  >
    branches.main_branch
  </span>
</span>
`;

exports[`#getTooltip should correctly format component information 1`] = `
"src/index.tsx

foo:src/index.tsx"
`;

exports[`#getTooltip should correctly format component information 2`] = `
"src/index.tsx

foo:src/index.tsx"
`;

exports[`#getTooltip should correctly format component information 3`] = `
"index.tsx

foo:src/index.tsx"
`;

exports[`#getTooltip should correctly format component information 4`] = `
"Foo

foo"
`;
